### PR TITLE
[SPARK-27971[SQL][R] MapPartitionsInRWithArrowExec.evaluate shouldn't eagerly read the first batch

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR is the same fix as https://github.com/apache/spark/pull/24816 but in vectorized `dapply` in SparkR.

## How was this patch tested?

Manually tested.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @BryanCutler and @mengxr

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 1]
```diff
@@ -243,28 +243,11 @@ case class MapPartitionsInRWithArrowExec(
...
-            false
-          }
-        }
-
-        override def next(): InternalRow = currentIter.next()
+      columnarBatchIter.flatMap { batch =>
+        val actualDataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType())
+        assert(outputTypes == actualDataTypes, "Invalid schema from dapply(): " +
+          s"expected ${outputTypes.mkString(", ")}, got ${actualDataTypes.mkString(", ")}")
+        batch.rowIterator.asScala
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: My guess is that it was wanting to check the data type at the first batch only. I think this fix is right.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
#24816 said `Issued fixed in #24734 but that PR might takes longer to merge.` and #24734 is merged yesterday. Do we still need this?

**Comment by HyukjinKwon** [comment]
Yea here's a different code path for R.

**Comment by HyukjinKwon** [comment]
Let me merge this one. R vectorization and Python vectorization code paths are being matched for the possibility to deduplicate both code path.

**Comment by felixcheung** [comment]
just curious, is it possible to share more infra between python and R so we don't constantly trying to match up with python afterwards?

**Comment by HyukjinKwon** [comment]
Yea .. I was thinking about that. I just wanted to like ugly deduplication ... let me probably try it in the near future with some neat deduplication. Let me cc you guys there.

**Pull request closed without merging**